docs/architecture/adr-060-baseapp++.md

```diff
@@ -0,0 +1,193 @@
+
+We define the general mempool interface as follows (subject to change):
+
+```go
+// MempoolTx we define a mempool transaction interface that is as minimal as
+// possible, only requiring applications to define the size of the transaction
+// to be used when reaping and getting the transaction itself. Interface type
+// casting can be used in the actual mempool implementation.
+type MempoolTx interface {
+	// Size returns the size of the transaction in bytes.
+	Size(codec.Codec) int
+	Tx() sdk.Tx
+}
+
+type Mempool[T MempoolTx] interface {
+	// Insert attempts to insert a MempoolTx into the mempool returning an error
+	// upon failure.
+	Insert(T) error
+	// Reap returns the next available transaction in the mempool, returning an error
+	// upon failure. The notion of 'available' or 'next' is defined by the application's
+	// mempool implementation.
+	Reap() (T, error)
+	// ReapMaxBytes performs the the same function as Reap, except it returns
+	// multiple transactions limited by their total size.
+	ReapMaxBytes(n int) ([]T, error)
+	// NumTxs returns the number of transactions currently in the mempool.
+	NumTxs() int
+}
+```
+
+We will define an implementation of `Mempool[T MempoolTx]` that will cover a
+majority of application use cases. Namely, it will prioritize transactions by
+priority and transaction sender, allowing for multiple prioritized transactions
+from the same sender. The mempool will be defined as a wrapper around a simple
+priority queue using a max binary heap, along with additional indexes/metadata
+to store senders and their nonces.
+
+Transaction reaping will essentially happen via a two-phase approach:
+
+1. Reap one or more transactions from the priority queue and collect them into a 
+   buffer.
+2. For each transaction in the buffer, ensure it does not violate any nonce predicates.
+3. Continue reaping until the desired number of valid transactions have been selected.
+4. Re-insert any transactions remaining in the buffer.
+
+```go
+type PriorityMempool[T MempoolTx] struct {
+	queue   *PriorityQueue[MempoolTx]
+	senders map[string][]int64
+	// ...
+}
+```
+
+Previous discussions<sup>1</sup> have come to the agreement that Tendermint will
+perform a request to the application, via `PrepareProposalRequest`, with a certain
+amount of transactions reaped from Tendermint's local mempool. The exact amount
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
+
+When Tendermint reaps transactions from the local mempool and sends them to the
+application via `PrepareProposalRequest`, the application will have to evaluate
+the transactions. Specifically, it will need to inform Tendermint if it should
+reject and or include each transaction. Note, the application can even _replace_
+transactions entirely with other transactions.
+
+When evaluating transactions from `PrepareProposalRequest`, the application will
```

Comment:
The basic mempool implementation will ignore them. You're encouraged to write your own mempool, which may or may not ignore them. The reason you'd want ignore them is so that the app can have full control over ordering, filtering, priority, etc.

Reply:
But now that I think of it...the design as proposed might not allow for this flexibility. I think we have to august the `Mempool` interface as such:

```go
type Mempool[T MempoolTx] interface {
  ReapMaxBytes(txRecords tendermint.TxRecords, maxBytes int) ([]T, error)
}
```

WDYT? Also, it probably makes no sense to just have `Reap`. We can probably remove that.

Reply:
That makes sense to me.